src/Documentation/Advanced-Concepts/Dependency-Injection.md

```diff
@@ -0,0 +1,75 @@
+---
+layout: page
+title: Dependency Injection
+---
+
+# What is Dependency Injection
+
+Dependency injection (DI) is a software design pattern that implements inversion of control for resolving dependencies.
+
+Orleans is using the abstraction written by the developers of [ASP.NET Core](https://docs.asp.net). For a detailed explanation about how does it work, check out the [official documentation](https://docs.asp.net/en/latest/fundamentals/dependency-injection.html#dependency-injection).
+
+# DI in Orleans
+
+Dependency Injection is currently supported only on the server side within Orleans.
+
+Orleans makes it possible to inject dependencies into application [Grains](../Getting-Started-With-Orleans/Grains.md).
+
+However Orleans supports every container dependent injection mechanisms, one of the most commonly used method is constructor injection.
+
+Theoretically any type can be injected which was previously registered in a [`IServiceCollection`](https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/Extensions/DependencyInjection/IServiceCollection/index.html) during Silo startup.
+
+Orleans also makes it possible to inject the following service into an application object instance:
+- `IGrainFactory`
+
+**Note**:
+As Orleans is evolving, as of the current plans it will be possible to leverage dependency injection in other application classes as well, like [`StreamProviders`](../Orleans-Streams/Stream-Providers.md). 
+
```

Comment:
I think there needs to be mention of the constructor of the grain being the place where the dependency is injected.  That wasn't immediately obvious to developers that are not versed in DI

## Constructors
Orleans still uses the notion of a class as the fundamental block for the grain.  It has a constructor which executed, but normally this constructor needs to be a empty default constructor with no arguments
Orleans uses the constructor to perform constructor injection (see Mark Seeman), where a given dependency is injected into the grain at(really before) activation time.
So to injection a necessary dependency into your grain, give the grain a constructor taking an argument
``` csharp
public MyGrain(ISuperCoolServiceClient client)
{
  this._client = client;
}
public async Task<string> FavoriteIceCream()
{
  return await client.MakeCallToDependency();
}
```
Where `ISuperCoolServiceClient` is a interface to the dependency you are injecting.   but it could be anything, a `int` constant, or a `string` for the name of your current pet